From 6c4be0eed15d0dc19e6fdfdefd67575a206c09a8 Mon Sep 17 00:00:00 2001
From: nibon7 <nibon7@163.com>
Date: Thu, 29 May 2025 19:00:56 +0800
Subject: [PATCH] shell: Enable digital signing action when document supports

When opening a comic and click "Sign Digitally...", papers crashes.

```console
thread 'main' panicked at shell/src/document_view/io.rs:522:14:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
---
 shell/src/document_view/actions.rs | 9 ++++++---
 shell/src/document_view/io.rs      | 2 +-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/shell/src/document_view/actions.rs b/shell/src/document_view/actions.rs
index 55e1455120fd..ef8cd936daba 100644
--- a/shell/src/document_view/actions.rs
+++ b/shell/src/document_view/actions.rs
@@ -67,9 +67,12 @@ impl imp::PpsDocumentView {
             self.set_action_enabled("add-text-annotation", false);
         }
 
-        if let Some(document) = document.dynamic_cast_ref::<DocumentSignatures>() {
-            self.set_action_enabled("digital-signing", document.can_sign());
-        }
+        let can_sign = document
+            .dynamic_cast_ref::<DocumentSignatures>()
+            .map(|doc| doc.can_sign())
+            .unwrap_or_default();
+
+        self.set_action_enabled("digital-signing", can_sign);
 
         self.set_action_enabled("dual-odd-left", dual_mode);
 
diff --git a/shell/src/document_view/io.rs b/shell/src/document_view/io.rs
index c58a97118bce..7834774eaaa7 100644
--- a/shell/src/document_view/io.rs
+++ b/shell/src/document_view/io.rs
@@ -519,7 +519,7 @@ impl imp::PpsDocumentView {
             .document()
             .and_dynamic_cast_ref::<DocumentSignatures>()
             .map(|d| d.available_signing_certificates())
-            .unwrap();
+            .unwrap_or_default();
         let dialog = adw::AlertDialog::new(Some(&gettext("Certificate Required")), None);
 
         if !certs.is_empty() {
-- 
2.49.0

